File: 445/c.py
n = int(input())
a = list(map(int, input().split()))
a = [i-1 for i in a]
ans = []
aaaaaa = 10**100


for i in range(n):
    tmp = a[i]
    time = 0
    sumi = set()
    if(a[i] == i):
        ans.append(a[i] + 1)
        continue
    while time <= n:
        if(tmp in sumi):
            break
        sumi.add(tmp)
        tmp = a[tmp]
        time += 1
    # tmp: ループの開始点
    # time: ループが始まるまでの時間
    # len(sumi)　- time: ループの長さ
    # (aaaaaa - time) % (len(sumi) - time) : ループの開始点から、aaaaaaまでの時間をループの長さで割った余り

    if( len(sumi) - time > 0):
        for i in range((aaaaaa - time) % (len(sumi) - time)):
            tmp = a[tmp]

    ans.append(a[tmp] +1)

print(*ans)

# for i in range(n):
#     tmp = a[i] - 1
#     time = 0
#     sumi = []

    # while True:
    #     # これを10^100回
    #     if(tmp in sumi and (not sumi)):
    #         break
    #     sumi.append(tmp)
    #     tmp = a[tmp] 
    #     time += 1
    #     print(sumi)
    # print(time,tmp,time-a.index(tmp))
    # time　- a.index(tmp) がループの長さ
    # (10^100 - a.index(tmp) ) % len(sumi) 

File: 445/c.cpp
#include <iostream>
#include <vector>
#include <cstring>
using namespace std;

int main() {
    int n;
    cin >> n;
    
    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
        a[i]--;  // 0-indexed
    }
    
    vector<int> ans(n);
    long long aaaaaa = 1;
    for (int i = 0; i < 100; i++) {
        aaaaaa *= 10;
    }
    
    for (int i = 0; i < n; i++) {
        int tmp = a[i];
        long long time = 0;
        vector<bool> sumi(n, false);
        
        // ループの開始点を見つける
        while (time <= n) {
            if (sumi[tmp]) {
                break;
            }
            sumi[tmp] = true;
            tmp = a[tmp];
            time++;
        }
        
        // tmp: ループの開始点
        // time: ループが始まるまでの時間
        // ループの長さを計算
        int loop_start = tmp;
        long long loop_length = 1;
        tmp = a[tmp];
        while (tmp != loop_start) {
            tmp = a[tmp];
            loop_length++;
        }
        
        // (aaaaaa - time) % loop_length 回移動
        if (loop_length > 0) {
            long long remaining = (aaaaaa - time) % loop_length;
            tmp = loop_start;
            for (long long j = 0; j < remaining; j++) {
                tmp = a[tmp];
            }
        }
        
        ans[i] = a[tmp] + 1;  // 1-indexed
    }
    
    // 結果を出力
    for (int i = 0; i < n; i++) {
        cout << ans[i];
        if (i < n - 1) cout << " ";
    }
    cout << endl;
    
    return 0;
}
